### PR TITLE
Implement swaps between underlying and PTs in marketplace

### PR DESCRIPTION
**Pull request by sourabhmarathe**
See #330. 

This PR also includes fixes to the mocked Pool.

### CODE REVIEW

File: build/marketplace/MarketPlace.sol [hunk 1]
```diff
@@ -22,11 +23,14 @@ contract MarketPlace {
...
-    /// markets are defined by a market pair which point to a fixed length array of principal token addresses.
-    /// the principal tokens those addresses represent correspond to their Principals enum value, thus the
-    /// array should be ordered in that way
+    /// markets are defined by a market pair which point to a fixed length array
+    /// of principal token addresses. The principal tokens those addresses
+    /// represent correspond to their Principals enum value, thus the array
+    /// should be ordered in that way
     mapping(address => mapping(uint256 => address[9])) public markets;
 
+    mapping(address => mapping(uint256 => address[9])) public pools;
```
Comment (robrobbins) [inline build/marketplace/MarketPlace.sol:32]: oh that looks wierd at first, but yeah, i guess there are 0-8 available "slots" for a pool that is associated with an individual principal in a market

i suppose then the pool can be referenced with the Principals enum the same way a given principal (in a market) can?

File: build/marketplace/MarketPlace.sol [hunk 2]
```diff
@@ -82,6 +86,91 @@ contract MarketPlace {
...
+        return true;
+    }
+
+    /// @notice sells the PT for the PT via the pool
+    /// @param p enum value of the principal token
+    /// @param u address of the underlying asset
+    /// @param m maturity (timestamp) of the market
+    /// @param a amount of PT to swap
+    /// @return uint128 amount of PT bought
+    function sellPT(
```
Comment (robrobbins) [inline build/marketplace/MarketPlace.sol]: so let's think about these names on a protocol big-picture. do we want `sellPT` or `sellPrincipalToken`

in our protocols generally (including swivel) we have gone with longer expressive method names

something to discuss

also gets us away from the `PT` vs `Pt` kind of argument...

File: build/marketplace/MarketPlace.sol [hunk 3]
```diff
@@ -82,6 +86,91 @@ contract MarketPlace {
...
+        return pool.buyPT(msg.sender, pool.buyPTPreview(a), a);
+    }
+
+    /// @notice sells the underlying for the PT via the pool
+    /// @param p enum value of the principal token
+    /// @param u address of the underlying asset
+    /// @param m maturity (timestamp) of the market
+    /// @param a amount of underlying to swap
+    /// @return uint128 amount of underlying sold
+    function sellUnderlying(
```
Comment (robrobbins) [inline build/marketplace/MarketPlace.sol:146]: this isn't `sellU` for example...

**Review by robrobbins — DISMISSED** [review]
let's discuss the names next week.

code looks good

File: build/marketplace/MarketPlace.sol (hunk 1 above)
Comment (sourabhmarathe) [reply build/marketplace/MarketPlace.sol:32]: Yes, that's what I was thinking.

**Review by robrobbins — APPROVED** [review]

**Pull request merged**